<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>Blast Tools &mdash; bio_utils 0.7.19a10 documentation</title>
    
    <link rel="stylesheet" href="_static/alabaster.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    './',
        VERSION:     '0.7.19a10',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <link rel="top" title="bio_utils 0.7.19a10 documentation" href="index.html" />
    <link rel="next" title="Contributing" href="contributing.html" />
    <link rel="prev" title="Verifiers" href="verifiers.html" />
   
  
  <meta name="viewport" content="width=device-width, initial-scale=0.9, maximum-scale=0.9" />

  </head>
  <body role="document">  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <div class="section" id="module-bio_utils.blast_tools">
<span id="blast-tools"></span><h1>Blast Tools<a class="headerlink" href="#module-bio_utils.blast_tools" title="Permalink to this headline">¶</a></h1>
<p>Package containing functions filtering BLAST output</p>
<p>Copyright:</p>
<blockquote>
<div><p>__init__.py functions filtering BLAST output
Copyright (C) 2015  William Brazelton, Alex Hyer</p>
<p>This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.</p>
<p>This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.</p>
<p>You should have received a copy of the GNU General Public License
along with this program.  If not, see &lt;<a class="reference external" href="http://www.gnu.org/licenses/">http://www.gnu.org/licenses/</a>&gt;.</p>
</div></blockquote>
<div class="section" id="introduction">
<h2>Introduction<a class="headerlink" href="#introduction" title="Permalink to this headline">¶</a></h2>
<p>The bio_utils&#8217; blast_tools subpackage contains a few tools for making BLAST
output easier to work with. They are from extensive but what they do provide
may be useful to some developers.</p>
</div>
<div class="section" id="blast-to-cigar">
<h2>blast_to_cigar<a class="headerlink" href="#blast-to-cigar" title="Permalink to this headline">¶</a></h2>
<p>A simple function that converts the query sequence, subject sequence, and
midline fields of BLAST+ XML output (M7) to a
<a class="reference external" href="http://drive5.com/usearch/manual/cigar.html">CIGAR string</a>. This function
supports both the newer and older versions of CIGAR strings.</p>
<dl class="function">
<dt id="bio_utils.blast_tools.blast_to_cigar">
<code class="descclassname">bio_utils.blast_tools.</code><code class="descname">blast_to_cigar</code><span class="sig-paren">(</span><em>query_seq</em>, <em>match_seq</em>, <em>subject_seq</em>, <em>cigar_age='new'</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/bio_utils/blast_tools/blast_to_cigar.html#blast_to_cigar"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#bio_utils.blast_tools.blast_to_cigar" title="Permalink to this definition">¶</a></dt>
<dd><p>converts BLAST alignment into a old or new CIGAR line</p>
<dl class="docutils">
<dt>Args:</dt>
<dd><p class="first">query_seq (str): Aligned portion of query sequence</p>
<p>match_seq (str): Alignment sequence</p>
<p>subject_seq (str): Aligned portion of subject/reference sequence</p>
<dl class="last docutils">
<dt>cigar_age (str): [&#8216;old&#8217;, &#8216;new&#8217;] CIGAR format to use, new is highly</dt>
<dd>detailed while old is fairly minimalistic</dd>
</dl>
</dd>
<dt>Returns:</dt>
<dd>str: CIGAR string</dd>
<dt>Raises:</dt>
<dd>ValueError: If query_seq, match_seq, and match_seq not same length</dd>
<dt>Examples:</dt>
<dd><div class="first last highlight-default"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">query</span> <span class="o">=</span> <span class="s">&#39;AAGGG--CCTTGTA&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">subject</span> <span class="o">=</span> <span class="s">&#39;AAGCCTTCCAGGTA&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignment_old</span> <span class="o">=</span> <span class="s">&#39;|||||  |||||||&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignment_new</span> <span class="o">=</span> <span class="s">&#39;AAG++  CC++GTA&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">blast_to_cigar</span><span class="p">(</span><span class="n">query</span><span class="p">,</span> <span class="n">alignment_new</span><span class="p">,</span> <span class="n">subject</span><span class="p">)</span>
<span class="go">3=2X2D2=2X3=</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">blast_to_cigar</span><span class="p">(</span><span class="n">query</span><span class="p">,</span> <span class="n">alignment_old</span><span class="p">,</span> <span class="n">subject</span><span class="p">,</span> <span class="n">cigar_age</span><span class="o">=</span><span class="s">&#39;old&#39;</span><span class="p">)</span>
<span class="go">5M2D7M</span>
</pre></div>
</div>
</dd>
</dl>
</dd></dl>

</div>
<div class="section" id="b6-evalue-filter">
<h2>b6_evalue_filter<a class="headerlink" href="#b6-evalue-filter" title="Permalink to this headline">¶</a></h2>
<p>This function iterates through any iterator yielding lines of a
<a class="reference internal" href="classes.html#b6entry"><span class="std std-ref">B6/M8</span></a> file. This iterator only returns the lines above an
E-value threshold as <a class="reference internal" href="classes.html#b6entry"><span class="std std-ref">B6Entry</span></a>.</p>
<dl class="function">
<dt id="bio_utils.blast_tools.b6_evalue_filter">
<code class="descclassname">bio_utils.blast_tools.</code><code class="descname">b6_evalue_filter</code><span class="sig-paren">(</span><em>handle</em>, <em>e_value</em>, <em>*args</em>, <em>**kwargs</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/bio_utils/blast_tools/filter_b6_evalue.html#b6_evalue_filter"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#bio_utils.blast_tools.b6_evalue_filter" title="Permalink to this definition">¶</a></dt>
<dd><p>Yields lines from handle with E-value less than or equal to e_value</p>
<dl class="docutils">
<dt>Args:</dt>
<dd><dl class="first docutils">
<dt>handle (file): B6/M8 file handle, can be any iterator so long as it</dt>
<dd>it returns subsequent &#8220;lines&#8221; of a B6/M8 entry</dd>
</dl>
<p>e_value (float): max E-value to return</p>
<p><a href="#id1"><span class="problematic" id="id2">*</span></a>args: Variable length argument list for b6_iter</p>
<p class="last"><a href="#id3"><span class="problematic" id="id4">**</span></a>kwargs: Arbitrary keyword arguments for b6_iter</p>
</dd>
<dt>Yields:</dt>
<dd>B6Entry: class containing all B6/M8 data</dd>
<dt>Example:</dt>
<dd><p class="first">Note: These doctests will not pass, examples are only in doctest
format as per convention. bio_utils uses pytests for testing.</p>
<div class="last highlight-default"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">b6_handle</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="s">&#39;test.b6&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">for</span> <span class="n">entry</span> <span class="ow">in</span> <span class="n">b6_evalue_filter</span><span class="p">(</span><span class="n">b6_handle</span><span class="p">,</span> <span class="mi">1</span><span class="n">e5</span><span class="p">)</span>
<span class="gp">... </span>    <span class="nb">print</span><span class="p">(</span><span class="n">entry</span><span class="o">.</span><span class="n">evalue</span><span class="p">)</span>  <span class="c"># Print E-value of filtered entry</span>
</pre></div>
</div>
</dd>
</dl>
</dd></dl>

</div>
<div class="section" id="query-sequence-retriever">
<h2>query_sequence_retriever<a class="headerlink" href="#query-sequence-retriever" title="Permalink to this headline">¶</a></h2>
<p>Retrieve the aligned query sequence for each alignment in a B6 file above an
E-value threshold.</p>
<dl class="function">
<dt id="bio_utils.blast_tools.query_sequence_retriever">
<code class="descclassname">bio_utils.blast_tools.</code><code class="descname">query_sequence_retriever</code><span class="sig-paren">(</span><em>fastaq_handle</em>, <em>b6_handle</em>, <em>e_value</em>, <em>fastaq='fasta'</em>, <em>*args</em>, <em>**kwargs</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/bio_utils/blast_tools/retrieve_query_sequences.html#query_sequence_retriever"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#bio_utils.blast_tools.query_sequence_retriever" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns FASTA entries for subject sequences from BLAST hits</p>
<p>Stores B6/M8 entries with E-Values below the e_value cutoff. Then iterates
through the FASTA file and if an entry matches the query of an B6/M8
entry, it&#8217;s sequence is extracted and returned as a FASTA entry
plus the E-Value.</p>
<dl class="docutils">
<dt>Args:</dt>
<dd><dl class="first docutils">
<dt>fastaq_handle (file): FASTA or FASTQ file handle, can technically</dt>
<dd>be any iterable that returns FASTA/Q &#8220;lines&#8221;</dd>
<dt>b6_handle (file): B6/M8 file handle, can technically</dt>
<dd>be any iterable that returns B6/M8&#8221;lines&#8221;</dd>
</dl>
<p>e_value (float): Max E-Value of entry to return</p>
<dl class="docutils">
<dt>fastaq (str): [&#8216;fasta&#8217;, &#8216;fastq&#8217;] whether file handle is a FASTA or</dt>
<dd>FASTQ file</dd>
</dl>
<p><a href="#id5"><span class="problematic" id="id6">*</span></a>args: Variable length argument list for b6_iter</p>
<p class="last"><a href="#id7"><span class="problematic" id="id8">**</span></a>kwargs: Arbitrary keyword arguments for b6_iter</p>
</dd>
<dt>Yields:</dt>
<dd><dl class="first last docutils">
<dt>FastaEntry: class containing all FASTA data</dt>
<dd>FastqEntry if fastaq=&#8217;fastq&#8217;</dd>
</dl>
</dd>
<dt>Example:</dt>
<dd><p class="first">Note: These doctests will not pass, examples are only in doctest
format as per convention. bio_utils uses pytests for testing.</p>
<div class="last highlight-default"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">fasta_handle</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="s">&#39;test.fasta&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">b6_handle</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="s">&#39;test.b6&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">for</span> <span class="n">entry</span> <span class="ow">in</span> <span class="n">query_sequence_retriever</span><span class="p">(</span><span class="n">fasta_handle</span><span class="p">,</span>
<span class="gp">... </span>                                      <span class="n">b6_handle</span><span class="p">,</span> <span class="mi">1</span><span class="n">e5</span><span class="p">)</span>
<span class="gp">... </span>    <span class="nb">print</span><span class="p">(</span><span class="n">entry</span><span class="o">.</span><span class="n">sequence</span><span class="p">)</span>  <span class="c"># Print aligned query sequence</span>
</pre></div>
</div>
</dd>
</dl>
</dd></dl>

</div>
<div class="section" id="subject-sequence-retriever">
<h2>subject_sequence_retriever<a class="headerlink" href="#subject-sequence-retriever" title="Permalink to this headline">¶</a></h2>
<p>Identical to <a class="reference internal" href="#query-sequence-retriever">query_sequence_retriever</a> except it returns subject sequences.</p>
<dl class="function">
<dt id="bio_utils.blast_tools.subject_sequence_retriever">
<code class="descclassname">bio_utils.blast_tools.</code><code class="descname">subject_sequence_retriever</code><span class="sig-paren">(</span><em>fasta_handle</em>, <em>b6_handle</em>, <em>e_value</em>, <em>*args</em>, <em>**kwargs</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/bio_utils/blast_tools/retrieve_subject_sequences.html#subject_sequence_retriever"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#bio_utils.blast_tools.subject_sequence_retriever" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns FASTA entries for subject sequences from BLAST hits</p>
<p>Stores B6/M8 entries with E-Values below the e_value cutoff. Then iterates
through the FASTA file and if an entry matches the subject of an B6/M8
entry, it&#8217;s sequence is extracted and returned as a FASTA entry
plus the E-Value.</p>
<dl class="docutils">
<dt>Args:</dt>
<dd><dl class="first docutils">
<dt>fasta_handle (file): FASTA file handle, can technically</dt>
<dd>be any iterable that returns FASTA &#8220;lines&#8221;</dd>
<dt>b6_handle (file): B6/M8 file handle, can technically</dt>
<dd>be any iterable that returns B6/M8 &#8220;lines&#8221;</dd>
</dl>
<p>e_value (float): Max E-Value of entry to return</p>
<p><a href="#id9"><span class="problematic" id="id10">*</span></a>args: Variable length argument list for b6_iter</p>
<p class="last"><a href="#id11"><span class="problematic" id="id12">**</span></a>kwargs: Arbitrary keyword arguments for b6_iter</p>
</dd>
<dt>Yields:</dt>
<dd>FastaEntry: class containing all FASTA data</dd>
<dt>Example:</dt>
<dd><p class="first">Note: These doctests will not pass, examples are only in doctest
format as per convention. bio_utils uses pytests for testing.</p>
<div class="last highlight-default"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">fasta_handle</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="s">&#39;test.fasta&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">b6_handle</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="s">&#39;test.b6&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">for</span> <span class="n">entry</span> <span class="ow">in</span> <span class="n">subject_sequence_retriever</span><span class="p">(</span><span class="n">fasta_handle</span><span class="p">,</span>
<span class="gp">... </span>                                      <span class="n">b6_handle</span><span class="p">,</span> <span class="mi">1</span><span class="n">e5</span><span class="p">)</span>
<span class="gp">... </span>    <span class="nb">print</span><span class="p">(</span><span class="n">entry</span><span class="o">.</span><span class="n">sequence</span><span class="p">)</span>  <span class="c"># Print aligned subject sequence</span>
</pre></div>
</div>
</dd>
</dl>
</dd></dl>

</div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
  <h3><a href="index.html">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">Blast Tools</a><ul>
<li><a class="reference internal" href="#introduction">Introduction</a></li>
<li><a class="reference internal" href="#blast-to-cigar">blast_to_cigar</a></li>
<li><a class="reference internal" href="#b6-evalue-filter">b6_evalue_filter</a></li>
<li><a class="reference internal" href="#query-sequence-retriever">query_sequence_retriever</a></li>
<li><a class="reference internal" href="#subject-sequence-retriever">subject_sequence_retriever</a></li>
</ul>
</li>
</ul>
<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="index.html">Documentation overview</a><ul>
      <li>Previous: <a href="verifiers.html" title="previous chapter">Verifiers</a></li>
      <li>Next: <a href="contributing.html" title="next chapter">Contributing</a></li>
  </ul></li>
</ul>
</div>
  <div role="note" aria-label="source link">
    <h3>This Page</h3>
    <ul class="this-page-menu">
      <li><a href="_sources/blast_tools.txt"
            rel="nofollow">Show Source</a></li>
    </ul>
   </div>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &copy;2015, William Brazelton, Alex Hyer.
      
      |
      Powered by <a href="http://sphinx-doc.org/">Sphinx 1.4.1</a>
      &amp; <a href="https://github.com/bitprophet/alabaster">Alabaster 0.7.7</a>
      
      |
      <a href="_sources/blast_tools.txt"
          rel="nofollow">Page source</a>
    </div>

    

    
  </body>
</html>